### PR TITLE
validate Docker digest and tag formatting

### PR DESCRIPTION
**Pull request by listx**
The unit tests are a little mundane, but they are better than nothing.

### CODE REVIEW

**Comment by listx** [comment]
This fixes #11.

**Comment by listx** [comment]
@bartsmykla can you take a look at this? Not sure how to add you as a reviewer because you are not in the GoogleCloudPlatform org... doh.

File: lib/dockerregistry/inventory.go [hunk 1]
```diff
@@ -119,9 +120,62 @@ func ParseManifestFromFile(filePath string) Manifest {
...
 
+// Validate checks for semantic errors in the yaml fields (the structure of the
+// yaml is checked during unmarshaling).
+func (m Manifest) Validate() error {
+	err := validateImages(m.Images)
+	if err != nil {
+		return err
+	}
+	return nil
+}
```
Comment (bartsmykla) [inline lib/dockerregistry/inventory.go]: Maybe:
```go
func (m Manifest) Validate() error {
	return validateImages(m.Images)
}
```
would be sufficient?

File: lib/dockerregistry/inventory.go [hunk 2]
```diff
@@ -119,9 +120,62 @@ func ParseManifestFromFile(filePath string) Manifest {
...
+
+func validateImages(images []Image) error {
+	var err error
+	for _, image := range images {
+		for digest, tagSlice := range image.Dmap {
+			err = validateDigest(digest)
+			if err != nil {
+				return err
+			}
```
Comment (bartsmykla) [inline lib/dockerregistry/inventory.go]: Maybe you could write:
```go
if err := validateDigest(digest); err != nil {
    return err
}
```

File: lib/dockerregistry/inventory.go [hunk 3]
```diff
@@ -119,9 +120,62 @@ func ParseManifestFromFile(filePath string) Manifest {
...
+
+			for _, tag := range tagSlice {
+				err = validateTag(tag)
+				if err != nil {
+					return err
+				}
```
Comment (bartsmykla) [inline lib/dockerregistry/inventory.go]: The same thing here:
```go
if err = validateTag(tag); err != nil {
    return err
}
```

File: lib/dockerregistry/inventory.go [hunk 4]
```diff
@@ -119,9 +120,62 @@ func ParseManifestFromFile(filePath string) Manifest {
...
+func (m Manifest) Validate() error {
+	err := validateImages(m.Images)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func validateImages(images []Image) error {
+	var err error
```
Comment (bartsmykla) [inline lib/dockerregistry/inventory.go]: If you'll refactor the error handling I commented below, you could get rid off this variable declaration.

File: lib/dockerregistry/inventory.go [hunk 5]
```diff
@@ -119,9 +120,62 @@ func ParseManifestFromFile(filePath string) Manifest {
...
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func validateDigest(digest Digest) error {
+	var validDigest = regexp.MustCompile(`^sha256:[0-9a-f]{64}$`)
```
Comment (bartsmykla) [inline lib/dockerregistry/inventory.go]: Why are you using here `var [...] =` and in other places `[...] :=`?

File: lib/dockerregistry/inventory.go [hunk 6]
```diff
@@ -119,9 +120,62 @@ func ParseManifestFromFile(filePath string) Manifest {
...
+	ok := validDigest.Match([]byte(digest))
+	if !ok {
+		return fmt.Errorf("invalid digest: %v", digest)
+	}
+	return nil
```
Comment (bartsmykla) [inline lib/dockerregistry/inventory.go]: I think I would refactor it to:
```go
if ok := validDigest.Match([]byte(digest)); !ok {
    return fmt.Errorf("invalid digest: %v", digest)
}
```

File: lib/dockerregistry/inventory.go [hunk 7]
```diff
@@ -119,9 +120,62 @@ func ParseManifestFromFile(filePath string) Manifest {
...
+}
+
+func validateTag(tag Tag) error {
+	var validTag = regexp.MustCompile(`^[\w][\w.-]{0,127}$`)
+	ok := validTag.Match([]byte(tag))
+	if !ok {
+		return fmt.Errorf("invalid tag: %v", tag)
+	}
```
Comment (bartsmykla) [inline lib/dockerregistry/inventory.go]: I think I would refactor it to:
```go
if ok := validTag.Match([]byte(tag)); !ok {
    return fmt.Errorf("invalid tag: %v", tag)
}
```

File: lib/dockerregistry/inventory.go [hunk 8]
```diff
@@ -119,9 +120,62 @@ func ParseManifestFromFile(filePath string) Manifest {
 func ParseManifest(bytes []byte) (Manifest, error) {
 	var m Manifest
 	err := yaml.UnmarshalStrict(bytes, &m)
+	if err != nil {
+		return m, err
+	}
+
+	err = m.Validate()
```
Comment (bartsmykla) [inline lib/dockerregistry/inventory.go]: Maybe you could write:
```go
if err := yaml.UnmarshalStrict(bytes, &m); err != nil {
    return m, err
}

return m, m.Validate()
```

File: lib/dockerregistry/inventory.go (hunk 1 above)
Comment (listx) [reply lib/dockerregistry/inventory.go]: Ack

**Comment by listx** [comment]
@bartsmykla Thx for the review!

The `if err := ...; err != nil` idiom is something I haven't gotten used to yet.

PTAL

**Comment by bartsmykla** [comment]
LGTM now :-)

**Pull request merged**